typed-lambda-calculus/article.md

```diff
@@ -0,0 +1,717 @@
+
+We then instantiate all placeholders we can:
+
+$$
+\begin{array}{c}
+\Gamma \vdash (\lambda x:Bool.\; x) : T_1 \to T_2;\quad \Gamma \vdash true : T_1 \\
+\hline
+\Gamma \vdash (\lambda x:Bool.\; x)\; true : T_2
+\end{array} \quad\text{(App)}
+$$
+
+For each condition, we do the same in turn:
+
+$$
+\begin{array}{c}
+\begin{array}{c}
+\begin{array}{c}
+x : T_2 \in \Gamma, x:Bool
+\\\hline
+\Gamma, x:Bool \vdash x : T_2
+\end{array}\quad\text{(Var)}
+\\\hline
+\Gamma \vdash (\lambda x:Bool.\; x) : Bool \to T_2;
+\end{array}\begin{array}{c}\\\text{(Abs)}\end{array}
+\quad
+\begin{array}{c}
+\\\\\Gamma \vdash true : Bool
+\end{array}\\
+\hline
+\Gamma \vdash (\lambda x:Bool.\; x)\; true : T_2
+\end{array}\quad\begin{array}{c}\\\\\text{(App)}\end{array}
+$$
+
+(notice how $T_1$ becomes $Bool$ due to the use of $\text{(Abs)}$ rule)
+
+And finally, we can go the other way, filling up the remaining placeholders:
+
+$$
+\begin{array}{c}
+\begin{array}{c}
+\begin{array}{c}
+x : Bool \in \Gamma, x:Bool
+\\\hline
+\Gamma, x:Bool \vdash x : Bool
+\end{array}\quad\text{(Var)}
+\\\hline
+\Gamma \vdash (\lambda x:Bool.\; x) : Bool \to Bool;
+\end{array}\begin{array}{c}\\\text{(Abs)}\end{array}
+\quad
+\begin{array}{c}
+\\\\\Gamma \vdash true : Bool \quad\text{(AxT)}
+\end{array}\\
+\hline
+\Gamma \vdash (\lambda x:Bool.\; x)\; true : Bool
+\end{array}\quad\begin{array}{c}\\\\\text{(App)}\end{array}
+$$
+
+### The impossibility of simply typed lambda calculus without atomic values
+
+So, we have a sound type system.
+However, it does have one pretty obvious drawback.
+
+<p><details>
+<summary>On the soundness</summary>
+
+We will omit the proof of the soundness of the simply typed lambda calculus with atomic booleans, however, it's easy enough to find for the interested reader, e.g. in (Pierce, 2002).
+
+The short sketch is to prove soundness we need to prove two propositions:
+
+1. A typed term can not be stuck.
+2. If a typed term is beta-reduced, the result is also a typed term.
+
+Both are proven by structural induction on terms, but you'll also have to formally define the computation semantics, which here we glossed over.
+
+***
+</details></p>
+
+In the untyped lambda calculus, we could define
+
+$$id = \lambda x.\; x$$
+
+and use it anywhere we would fancy an identity function.
+
+With our new fancy type system, we can not.
+Obviously, for any concrete type, we could define
+
+$$id_{Bool} = \lambda x : Bool.\; x,$$
+$$id_{Bool \to Bool} = \lambda x : Bool \to Bool.\; x,$$
+$$\ldots$$
+
+But we can't do it in general.
+
+Notice how we need a terminating type in the type signature.
+Consequently, if our calculus didn't have any atomic values, i.e. didn't have any types beyond abstractions, we wouldn't be able to specify the argument type.
+
+The reason for this is the lack of polymorphism.
+
+## Polymorphic lambda-calculi
+
+Polymorphism, essentially, is the ability of the type system to represent terms that contain arbitrary types.
```

Comment:
In what sense do terms "contain" types? Maybe you're looking for a different verb